[Woo POS] Coupons: Keep products and coupons scroll state separate (Using tab views)

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Products and Coupons reused a single list, which shared the same scrolling state.

https://github.com/woocommerce/woocommerce-ios/pull/15536 PR attempted to fix this issue by saving the current scroll state for a given item type. However, testing has revealed that the behavior is not entirely stable.

## Solution

I propose an alternative approach of using SwiftUI `TabView` to achieve the same behavior. Passing different lists for different `ItemListType` ensures that their scroll state are independent. 

### Limitations

This solution, however, doesn't preserve the item list state at all times:
- When we switch to checkout and back, the whole `ItemListView` is removed, resulting in lost state
- When we enter search input, `InputListView` is also reloaded, resulting in a reset scroll state for all the "tabs". I tried to investigate that, but I couldn't find a quick solution. My conclusion is that we need to sort out how we handle item and container states between `PointOfSaleDashboardView` and `ItemListView`, since entering `searchText` or changing the `selectedItemListType` also updates the dashboard, which cascades the updates. 
- We also remove and add `ItemList` (and therefore update state) when switching between recent searches and item list, but this is expected behavior for search.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Test with feature flags disabled, confirm that the item list behaves as before
- Enable search and coupons feature flags, confirm that coupons and products preserve their own scroll states

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

Tested on iPad Air 11 Inch M2 Simulator with various feature flag combinations

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### General Behavior
https://github.com/user-attachments/assets/dab48aa0-f4ee-4f73-a807-aad7c53e0522

### Search-Only

https://github.com/user-attachments/assets/c80532ec-4c66-4e14-9982-7c9fbdb7a38a

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Maintain item list scroll state by using tab view

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -84,26 +84,20 @@ struct ItemListView: View {
+                    .gesture(DragGesture())
+                itemListContent
+                    .tag(ItemListType.products(search: true))
+                    .gesture(DragGesture())
```
The way to disable scrolling between the pages

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -122,6 +116,30 @@ struct ItemListView: View {
     }
+
+    @ViewBuilder
+    var itemListContent: some View {
```
Moving `ItemList` together with `POSRecentSearchesView` so that switching between the two wouldn't clear the date as well.

iamgabrielma commented:
This works great, as mentioned in https://github.com/woocommerce/woocommerce-ios/pull/15536#issuecomment-2822865166 I think has more benefits that the initial approach and we should go ahead with this solution. It's a great use of the TabView that I didn't know we could use like this.

I've tested  on simulator iPad mini A17 Pro - iOS 18.3 with both coupons (ofc 😂 ) and search flags on and I saw no issues, quite the opposite. I'll mark this one as ready for review and pre-approve it in case you want to merge it directly.

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -84,26 +84,20 @@ struct ItemListView: View {
-            }
+            TabView(selection: $selectedItemListType) {
+                itemListContent
+                    .tag(ItemListType.products(search: false))
```
I didn't realize that `.tag` accepts a generic type, I though it needed to be a String/Int 💯

staskus commented:
@iamgabrielma, thanks for being proactive and reviewing it. I will finalize and retest a bit more to make sure it works fine before merging. 👍

staskus committed:
Merge branch 'trunk' into task/woomob-322-separate-item-type-scroll-state-2

joshheald reviewed (commented):
I also like this approach; just based on playing with it a little, but it doesn't change anything about the state management, which is nice.

The drag gesture stuff is interesting though... it enables search when you drag. I guess that's because it's the hidden middle tab, and you can drag between the types?

https://github.com/user-attachments/assets/9195b764-0d1e-4fe9-b9c6-70a50c19c3ca

Taking the `gesture` off the search tab doesn't fix it though. I wonder whether we should actually just drag between the base types, without the search value involved at all. Search being a tab doesn't fit the model of it; we have a products tab, which has a search mode.

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -84,26 +84,20 @@ struct ItemListView: View {
+                itemListContent
+                    .tag(ItemListType.products(search: false))
+                    .gesture(DragGesture())
+                itemListContent
+                    .tag(ItemListType.products(search: true))
+                    .gesture(DragGesture())
```
If you remove these three lines, I think the search/drag issue is solved.

You can't drag from search to other tabs (which makes sense) and you don't activate search by dragging.

staskus replied:
Could you clarify this point a bit?

Does the dragging gesture work for you? I added `.gesture(DragGesture())` to remove the default gesture.

>  I wonder whether we should actually just drag between the base types, without the search value involved at all. 

I went back and forth on this. We could. The downside is that we would reuse the same list for products and product search. Which behavior makes sense to you, @joshheald? Also, please take a look at the updated description.

staskus replied:
It could be related to `DashboardView` holding the state of selectedItem and search term. Within the body, we're switching over `itemsViewState.containerState`. When we enter a search term, the `itemViewState` changes to loading and then to `content`. Maybe these changes cause the redraw. When we simply change the selected item, these kind of state changes are not happening.

staskus replied:
I think DashboardView shouldn't be dependent on `selectedItemListType`, this is a state of the specific item type, which is handled within `ItemListView`.

staskus replied:
I'll experiment with moving `headerView` and `tabView` to `DashboardView`. It's easier than the reverse, since we would need to refactor how we handle container states.

joshheald replied:
Sorry, I didn't see your earlier comment for some reason – the notification only came through a moment ago.

DashboardView mainly depends on it for historical reasons of the container state... if that were separated out, or we stopped doing fullscreen loading/errors, it could perhaps not be referenced there. 

But it still needs to be higher up the stack than ItemListView so that it can survive going back and forth from the cart.

staskus committed:
Optimize ItemList TabView solution to prevent updates when selectedItemListType changes

Configure ItemList separately for each ItemListType.

It means turning itemsController and itemListState from vars that use selectedItemListType to func that accept an item list type.

Each ItemList within TabView can be now configured with different itemListType. The views won't be updating when selectedItemListType changes and the state is preserved.

staskus committed:
Add product search and coupons lists depending on the feature flag

staskus committed:
Merge branch 'trunk' into task/woomob-322-separate-item-type-scroll-state-2

staskus commented:
@joshheald @iamgabrielma I made a few iterations internally until I settled on this. I'd appreciate a second review and testing round. 🙇 

There are a few limitations, mainly:

- List scroll states reset when the search input is entered
- List scroll states reset when checkout is done

In other scenarios, the list scroll states for Products and Coupons should be independent and preserved while switching.

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -81,42 +68,64 @@ struct ItemListView: View {
-            if shouldShowSearchField && searchTerm.isEmpty {
+            TabView(selection: $selectedItemListType) {
+                itemListContent(.products(search: false))
+                if isSearchProductsFeatureEnabled {
+                    itemListContent(.products(search: true))
+                }
```
I don't really understand why the unsearched list's scroll state changes when we enter a search term, given that search is another tab... can you explain why changing to this tab doesn't reset the scroll state, but when you enter a search it does?

Do you think search even should be a tab?

staskus replied:
> can you explain why changing to this tab doesn't reset the scroll state, but when you enter a search, it does?

No. I couldn't identify the exact state change that causes the redraw. My understanding is that the whole `ItemListView` is redrawn when `searchTerm` changes, and all the tabs as well. Why search term change causes it I couldn't understand yet.

> Do you think search even should be a tab?

Not necessarily. I explained in the comment tagging for the review. We probably shouldn't reuse the same list for search and product list since it could suffer from the same issue - shared scroll state. Then we could choose to have different lists:
- Use `switch` statement for search and non-search versions, which would result in a new view created every time we open and quit search, therefore the product list scroll state wouldn't be maintained in-between search sessions
- Use `TabView` within `TabView` (or some other similar solution) to maintain a scroll state for the product list in-between searches

joshheald replied:
I'll see if I can spot it.

joshheald replied:
I've had a quick look, and it's nothing obvious.

My suspicion is that there's just too much state that each bit of ItemListView depends on, or appears to.

I think SwiftUI is much better at optimising view updates when things are divided into views, not just viewbuilders; at least, that's what I found when we moved over to Observable.

Tomorrow I'll try factoring some of the viewbuilders out of ItemListView into full views, to see if that helps.

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -81,42 +68,64 @@ struct ItemListView: View {
-            }
-        })
+        .tag(itemListType)
+        .gesture(DragGesture()) // Disable a default swipe gesture between the tabs
```
That's a strange looking way to disable a gesture! thanks for the comment.

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -13,40 +13,27 @@ struct ItemListView: View {
 
     private var isAddingCouponAllowed: Bool {
         guard case .coupons = selectedItemListType else { return false }
+        let itemListState = itemListState(selectedItemListType)
```
I can't find any problems with this state indirection, but I have the fear! I've tested things like variations, adding a coupon, PTR, load more, empty state, it all seems fine.

staskus replied:
I don't think you should be fearful. In this case, the behavior is identical; we take an itemListState of `selectedItemListType`.

joshheald replied:
Yeah, it's whether the change observation continues to work correctly that was worrying me. But clearly, it works fine!

joshheald reviewed (commented):
Looks good so far, I want to check it with the new view state coordinator approach though.

It would be great to keep search state when going back and forth from the cart, and especially after a search. What are the issues preventing that, do you think?

staskus commented:
> It would be great to keep search state when going back and forth from the cart, and especially after a search. What are the issues preventing that, do you think?

@joshheald, `PointOfSaleDashboardView` `contentView` has an if statement, which removes `ItemListVIew` when we switch from `.building` to -> `.finalizing`. That's my understanding.

joshheald reviewed (approved):
Yep, I've tested with the other state change merged and it's still working as described here.

I think we could usefully add the top item for each list to the view state coordinator, if that helps keep the scroll position even more. Another candidate would be the navigation path, so that we can come back to a variation list when editing the order from the checkout as well.

staskus committed:
Merge branch 'trunk' into task/woomob-322-separate-item-type-scroll-state-2

staskus commented:
I'm merging this PR.

I tried some more solutions with splitting into smaller SwiftUI structs and making changes around the dashboard but haven't discovered what works.

This is still an improvement:
- Products and Coupons don't share states
- Products and Coupons retain a scroll state when we switch between the tabs
